lib/bundling.js

```diff
@@ -0,0 +1,326 @@
+  for (var i = 0; i < discriminatorFields.length; ++i) {
+    var id = _.at(obj, discriminatorFields[i])[0];
+    if(id === undefined) {
+      return undefined;
+    }
+    ids.push(id);
+  }
+  return ids.join(',');
+}
+exports.computeBundleId = computeBundleId;
+
+/**
+ * A task coordinates the execution of a single bundle.
+ *
+ * @param {function} apiCall - The function to conduct calling API.
+ * @param {Object} bundlingRequest - The base request object to be used
+ *   for the actual API call.
+ * @param {String} bundledField - The name of the field in bundlingRequest
+ *   to be bundled.
+ * @param {?String} subresponseField - The name of the field in the response
+ *   to be passed to the callback.
+ * @constructor
+ * @private
+ */
+function Task(apiCall, bundlingRequest, bundledField, subresponseField,
+              byteLengthFunction) {
+  this._apiCall = apiCall;
+  this._request = bundlingRequest;
+  this._bundledField = bundledField;
+  this._subresponseField = subresponseField;
+  this._byteLengthFunction = byteLengthFunction;
+  this._data = [];
+}
+
+exports.Task = Task;
+
+/**
+ * Returns the number of elements in a task.
+ */
+Task.prototype.getElementCount = function() {
+  var count = 0;
+  for (var i = 0; i < this._data.length; ++i) {
+    count += this._data[i].elements.length;
+  }
+  return count;
+};
+
+/**
+ * Returns the total byte size of the elements in a task.
+ */
+Task.prototype.getRequestByteSize = function() {
+  var size = 0;
+  for (var i = 0; i < this._data.length; ++i) {
+    size += this._data[i].bytes;
+  }
+  return size;
+};
+
+Task.prototype.getEventEmitters = function() {
+  var emitters = [];
+  for (var i = 0; i < this._data.length; ++i) {
+    if (this._data[i].emitter) {
+      emitters.push(this._data[i].emitter);
+    }
+  }
+  return emitters;
+};
+
+/**
+ * Invokes the actual API call with current elements.
+ */
+Task.prototype.run = function() {
+  if (this._data.length === 0) {
+    return;
+  }
+  var request = this._request;
+  var elements = [];
+  for (var i = 0; i < this._data.length; ++i) {
+    elements.push.apply(elements, this._data[i].elements);
+  }
+  request[this._bundledField] = elements;
+
+  this._apiCall(request, function(err, response) {
+    var subresponses = [];
+    if (err || !this._subresponseField) {
+      this._data.forEach(function() { subresponses.push(null); });
+    } else {
+      var responseValue = response[this._subresponseField];
+      var index = 0;
+      this._data.forEach(function(data) {
+        var endIndex = index + data.elements.length;
+        if (endIndex > responseValue.length) {
+          console.warn('subresponse size is not enough.');
+          return;
+        }
+        subresponses.push(responseValue.slice(index, endIndex));
+        index = endIndex;
+      });
+    }
+    for (var i = 0; i < this._data.length; ++i) {
+      if (this._data[i].emitter) {
+        this._data[i].emitter._callback(err, subresponses[i]);
```

Comment:
As discussed offline, let's actually pass the entire return object, with the object's subresponse field set to only contain `subresponses[i]`. (See: https://github.com/googleapis/gax-python/blob/master/google/gax/bundling.py#L194)